agave_app/GLView3D.cpp

```diff
@@ -287,41 +287,53 @@ GLView3D::FitToScene()
 }
 
 void
-GLView3D::keyPressEvent(QKeyEvent* event)
+GLView3D::toggleAreaLightRotateControls()
 {
-  static enum MODE { NONE, ROT, TRANS } mode = MODE::NONE;
+  // toggle rotate tool
+  if (m_areaLightMode == AREALIGHT_MODE::NONE || m_areaLightMode == AREALIGHT_MODE::TRANS) {
+    m_viewerWindow->showAreaLightGizmo(true);
+    m_viewerWindow->setTool(
+      new RotateTool(m_viewerWindow->m_toolsUseLocalSpace, ManipulationTool::s_manipulatorSize * devicePixelRatioF()));
+    m_viewerWindow->forEachTool(
+      [this](ManipulationTool* tool) { tool->setUseLocalSpace(m_viewerWindow->m_toolsUseLocalSpace); });
+    m_areaLightMode = AREALIGHT_MODE::ROT;
+  } else {
+    m_viewerWindow->showAreaLightGizmo(false);
+    m_viewerWindow->setTool(nullptr);
+    m_areaLightMode = AREALIGHT_MODE::NONE;
+  }
+}
 
+// TODO currently this function is not wired up to any gui at all.
+// This is because translation of area light source still needs work.
+// (Currently rotation is sufficient.)
+void
+GLView3D::toggleAreaLightTranslateControls()
+{
+  // toggle translate tool
+  if (m_areaLightMode == AREALIGHT_MODE::NONE || m_areaLightMode == AREALIGHT_MODE::ROT) {
+    m_viewerWindow->showAreaLightGizmo(true);
+    m_viewerWindow->setTool(
+      new MoveTool(m_viewerWindow->m_toolsUseLocalSpace, ManipulationTool::s_manipulatorSize * devicePixelRatioF()));
+    m_viewerWindow->forEachTool(
+      [this](ManipulationTool* tool) { tool->setUseLocalSpace(m_viewerWindow->m_toolsUseLocalSpace); });
+    m_areaLightMode = AREALIGHT_MODE::TRANS;
+  } else {
+    m_viewerWindow->showAreaLightGizmo(false);
+    m_viewerWindow->setTool(nullptr);
+    m_areaLightMode = AREALIGHT_MODE::NONE;
+  }
+}

```

Comment:
A note for the future, there's some duplicated boilerplate between these two methods. Might be worth seeing if they can be combined or reduced whenever the translation controls go in.

Reply:
Yes, that's something for later. There's a big TODO at the top of the translate function and it may be completely changed when it's time to wire it up to something.  This code is here to sort of remind us that it's similar to the rotate code but they can work together with the mode variable. Think I should just delete it?

Reply:
I think it's fine to keep in, it's a good reminder.